icx-proxy/src/main.rs

```diff
@@ -216,25 +222,55 @@ async fn forward_request(
     }
 
     let canister = HttpRequestCanister::create(agent.as_ref(), canister_id);
-    let result = canister
-        .http_request(method, uri.to_string(), headers, &entire_body)
+    let query_result = canister
+        .http_request(
+            method.clone(),
+            uri.to_string(),
+            headers.clone(),
+            &entire_body,
+        )
         .call()
         .await;
 
-    // If the result is a Replica error, returns the 500 code and message. There is no information
-    // leak here because a user could use `dfx` to get the same reply.
-    let (http_response,) = match result {
-        Ok(response) => response,
-        Err(AgentError::ReplicaError {
-            reject_code,
-            reject_message,
-        }) => {
-            return Ok(Response::builder()
+    fn handle_result(
+        result: Result<(HttpResponse,), AgentError>,
+    ) -> Result<HttpResponse, Result<Response<Body>, Box<dyn Error>>> {
+        // If the result is a Replica error, returns the 500 code and message. There is no information
+        // leak here because a user could use `dfx` to get the same reply.
+        match result {
+            Ok((http_response,)) => Ok(http_response),
+            Err(AgentError::ReplicaError {
+                reject_code,
+                reject_message,
+            }) => Err(Ok(Response::builder()
                 .status(StatusCode::INTERNAL_SERVER_ERROR)
                 .body(format!(r#"Replica Error ({}): "{}""#, reject_code, reject_message).into())
-                .unwrap());
+                .unwrap())),
+            Err(e) => Err(Err(e.into())),
         }
-        Err(e) => return Err(e.into()),
+    }
+
+    let http_response = match handle_result(query_result) {
+        Ok(http_response) => http_response,
+        Err(response_or_error) => return response_or_error,
+    };
+
+    let http_response = if http_response.upgrade {
+        let waiter = garcon::Delay::builder()
+            .throttle(std::time::Duration::from_millis(500))
+            .timeout(std::time::Duration::from_secs(60 * 5))
```

Comment:
> Update calls require polling. Is there a timeout here, so that slow canisters who never respond don't keep the proxy busy?

@nomeata does this answer that?

Reply:
FWIW, I think I copied this timeout from somewhere else and forgot to bring it up. 5 minutes seems like a long time.

Reply:
It looks like it answers the question, unless the libraries are misleading.

Also it makes a big difference if this is threaded and threads are cheap and a long polling request is not actually expensive, or if we have to worry about resources here.

Reply:
5 minutes is far too long.  15 seconds is plenty IMO

Reply:
Updated to 15 seconds in https://github.com/dfinity/agent-rs/pull/195/commits/a1ec3a4548b58bebe19164cb7e272da7b80a7f2c